app/controllers/nodes_controller.rb

```diff
@@ -26,4 +26,29 @@ def get_all_nodes
 
     render json: matching_nodes, root: 'data', each_serializer: GetAllNodesSerializer
   end
+
+  def get_node_attributes
+
+    start_year = params[:start_year].to_i
+    end_year = params[:end_year].to_i
+
+    unless params[:start_year].present?
+      raise ActionController::ParameterMissing, 'Required start_year missing'
+    end
+
+    unless params[:end_year].present?
+      raise ActionController::ParameterMissing, 'Required end_year missing'
+    end
+
+    node_ids = Node
+                   .select('node_id')
+                   .joins('LEFT JOIN context_nodes ON nodes.node_type_id = context_nodes.node_type_id')
+                   .where('context_nodes.context_id = ?', @context.id)
+
+    ind_values = NodeInd.get_attributes_for_nodes(node_ids, @context.id, start_year, end_year)
```

Comment:
I don't like this approach of first retrieving the node ids and then putting them in an enormous IN clause - such queries tend to be slow. Could we try what happens if we changed get_attributes_for_nodes into get_attributes_for_context, so that the queries in NodeInd and NodeQuant perform an inner join with context_nodes instead?

Reply:
:+1: